### PR TITLE
Implement AVL tree

### LINKED ISSUE #16: Implement AVL Tree
With 2 basic operations:
- Insert a value into AVL Tree: `O(log n)`
- Remove a value from AVL Tree: `O(log n)`

### PR DESCRIPTION
**Pull request by hellovietduc**
Closes: #16

### CODE REVIEW

**Comment by hellovietduc** [comment]
I implemented AVL tree in a package called `avl` by writing Binary Search Tree first (package named `bst`), and then imported `bst` like this:

```go
import (
	"github.com/vietduc01100001/dna-go/bst"
)
```

But since I don't code on the `master` branch, `go get` can't download the package and the compiler don't know where the the package `bst` comes from.

How do I import `bst` in package `avl` in this situation? @thanks173

**Comment by thanks173** [comment]
> How do I import `bst` in package `avl` in this situation? @thanks173

First, just run `go get github.com/vietduc01100001/dna-go/bst`
Then go to your package folder in `$GOPATH/src/github.com/vietduc01100001/dna-go/` and checkout to your desired branch.

File: avl/avl.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+			n.Right = newNode
+		} else {
+			n.Right.Insert(value)
+			n.rightHeight++
+			n.ensureBalance()
+		}
+	}
+}
+
+func (n *TreeNode) ensureBalance(node *Node) {
```
Comment (thanks173) [inline avl/avl.go]: what is this `node` and its struct?

**Review by thanks173 — CHANGES_REQUESTED** [review]
After the rotation, the tree root could be changed. How do you keep track of that?

Also, your bst implementation contains bug, leaving a duplicate node after deletion.

File: bst/bst.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		} else {
+			parent.Right = nil
+		}
+		return
+	}
+
+	if node.Left != nil {
+		maxVal := node.Left.FindMax()
+		node.Value = maxVal
+		node.Left.Delete(value)
```
Comment (thanks173) [inline bst/bst.go]: there isn't a node with `value` in the left subtree.

File: bst/bst.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		return
+	}
+
+	if node.Right != nil {
+		minVal := node.Right.FindMin()
+		node.Value = minVal
+		node.Right.Delete(value)
```
Comment (thanks173) [inline bst/bst.go]: there isn't a node with `value` in the right subtree.

**Comment by hellovietduc** [comment]
> First, just run `go get github.com/vietduc01100001/dna-go/bst`
> Then go to your package folder in `$GOPATH/src/github.com/vietduc01100001/dna-go/` and checkout to your desired branch.

But `bst` isn't in the `master` branch, and as I know, `go get` only gets from the `master` branch.

**Comment by thanks173** [comment]
> 
> But `bst` isn't in the `master` branch, and as I know, `go get` only gets from the `master` branch.

`go get` is a wrapper around `git clone`. `$GOPATH/src/github.com/vietduc01100001/dna-go/` is your project folder and you can checkout this folder to any branch you want. You also can (and should) use this location to code.

**Comment by hellovietduc** [comment]
Miss clicked, sorry!

**Pull request closed without merging**